File: Matchismo OBJC/Model/CardMatchingGame.m
//
//  CardMatchingGame.m
//  Matchismo OBJC
//
//  Created by Пермяков Андрей on 6/22/19.
//  Copyright © 2019 Пермяков Андрей. All rights reserved.
//

#import "CardMatchingGame.h"

@interface  CardMatchingGame()
@property (nonatomic, readwrite) NSInteger score;
@property (nonatomic, strong) NSMutableArray *cards;
@end

@implementation CardMatchingGame

- (NSMutableArray *)cards {
    if (!_cards) _cards = [[NSMutableArray alloc] init];
    return _cards;
}

- (instancetype)initWithCardCount:(NSUInteger)count
                        usingDeck:(Deck *)deck {
    self = [super init];
    
    if (self) {
        for (int i = 0; i < count; i++) {
            Card *card = [deck drawRandomCard];
            if (card) {
                [self.cards addObject:card];
            } else {
                self = nil;
                break;
            }
        }
    }
    
    return self;
}

- (Card *)cardAtIndex:(NSUInteger)index {
    return (index < [self.cards count]) ? self.cards[index] : nil;
}

static const int MISMATCH_PENALTY = 2;
static const int MATCH_BONUS = 4;
static const int COST_TO_CHOOSE = 1;


- (void)chooseCardAtIndex:(NSUInteger)index {
    Card *card = [self cardAtIndex:index];
    if (!card.isMatched) {
        if (card.isChosen) {
            card.chosen =  NO;
        } else {
            NSMutableArray *otherCards = [[NSMutableArray alloc] init];
            for (Card *otherCard in self.cards) {
                if (!otherCard.isMatched && otherCard.isChosen) {
                    [otherCards addObject:otherCard];
                    if ([otherCards count] == self.numberOfCardMatchingMode - 1) {
                        int matchScore = [card match: otherCards];
                        if (matchScore) {
                            self.score += matchScore * MATCH_BONUS;
                            card.matched = YES;
                            [self markMatchedCards:otherCards];
                        } else {
                            self.score -= MISMATCH_PENALTY * (pow(self.numberOfCardMatchingMode - 1, 3)); //bigger mismatch penalty when mismatched more cards. If we got a 2 card game,pow(1, n) returns 1
                            [self unchooseCards:otherCards];
                        }
                        break;
                    }
                }
            }
            self.score -= COST_TO_CHOOSE * (self.numberOfCardMatchingMode - 1); // again, the cost is bigger when awards are bigger
            card.chosen = YES;
        }
    }
}

- (void)unchooseCards:(NSArray *)cards {
    for (Card *card in cards) {
        card.chosen = NO;
    }
}

- (void)markMatchedCards:(NSArray *)cards {
    for (Card *card in cards) {
        card.matched = YES;
    }
}

@end

File: Matchismo Swift/Model/CardMatchingGame.swift
//  Converted to Swift 5 by Swiftify v5.0.37171 - https://objectivec2swift.com/
//
//  CardMatchingGame.swift
//  Matchismo
//
//  Created by Пермяков Андрей on 6/22/19.
//  Copyright © 2019 Пермяков Андрей. All rights reserved.
//

import Foundation

class CardMatchingGame: NSObject {
  private(set) var score = 0
  var numberOfCardMatchingMode = 0
  private(set) var matchDesctiption = ""
  private var cards: [Card]
  
  // Designated initializer
  init?(cardCount count: Int, using deck: Deck) {
    cards = []
    for _ in 0..<count {
      let card = deck.drawRandomCard()
      guard card != nil else {
        return nil
      }
      // Use non-optional `card` further in the code
      cards.append(card!)
    }
  }
  
  func chooseCard(at index: Int) {
    guard let card = self.card(at: index), !card.isMatched else { return }
    matchDesctiption = ""
    if card.isChosen {
      card.isChosen = false
      return
    }
    var otherCards: [Card] = []
    for otherCard in cards {
      if !otherCard.isMatched && otherCard.isChosen {
        otherCards.append(otherCard)
        // We got an arr otherCards + current card involved in a match
        if (otherCards.count + 1) == numberOfCardMatchingMode {
          var matchScore = card.match(otherCards)
          modifyScoreInformation(&matchScore, withCards: otherCards + [card])
          break
        }
      }
    }
    score -= ScoreMultipliers.costToChoose * (numberOfCardMatchingMode - 1)
    card.isChosen = true
  }
  
  private func modifyScoreInformation(_ matchScore: inout Int, withCards cards: [Card]) {
    let cardsDescription = cardsContents(cards)
    if matchScore != 0 {
      matchScore *= ScoreMultipliers.matchBonus
      score += matchScore
      markMatchedCards(cards)
      matchDesctiption = "Matched \(cardsDescription) for \(matchScore) points"
    } else {
      matchScore = -(ScoreMultipliers.mismatchPenalty *
        Int((pow(Double(numberOfCardMatchingMode), 3))))
      score += matchScore
      unchooseCards(cards)
      matchDesctiption = "\(cardsDescription) is a mismatch! \(matchScore) penalty"
    }
  }
  
  func card(at index: Int) -> Card? {
    return (index < cards.count) ? cards[index] : nil
  }
  
  private func cardsContents(_ cards: [Card]) -> String {
    var contents = ""
    for card in cards {
      contents += (card.contents)
    }
    return contents
  }
  
  private func unchooseCards(_ cards: [Card]) {
    for card in cards {
      card.isChosen = false
    }
  }
  
  private func markMatchedCards(_ cards: [Card]) {
    for card in cards {
      card.isMatched = true
    }
  }
}

struct ScoreMultipliers {
  static let mismatchPenalty = 2
  static let matchBonus = 4
  static let costToChoose = 1
}
